### ISSUE TITLE
ModelCheckpoint tries to find monitored key on non-zero ranks

### ISSUE THREAD

**Issue by gshaikov** [Lightning-AI/lightning#15852]
### Bug description

Hi, we're using `ModelCheckpoint` with `save_top_k=2` to keep track of the best checkpoints based on `monitor=val_auc`.

We're computing and logging `val_auc` on `validation_epoch_end`, however we only do it on *rank zero*.

```python
def validation_epoch_end(self, outputs):
    gathered_outputs = gloo_gather(outputs)
    if rank_zero():
        auc = compute_auc(gathered_outputs)
        self.log('val_auc': auc)
```

If we run this model over >1 GPUs, we get an error:
```
pytorch_lightning.utilities.exceptions.MisconfigurationException: `ModelCheckpoint(monitor='val_auc')` could not find the monitored key in the returned metrics: ['train_loss', 'val_loss', 'epoch', 'step']. HINT: Did you call `log('val_auc, value)` in the `LightningModule`?
```

This tells us that `ModelCheckpoint` ran on other ranks too, not just on rank zero. This looks like a wrong behaviour - model checkpoint shouldn't run on non-zero ranks, in our understanding.

Is our configuration wrong or is this a bug? Thanks.

Related error was posted already in discussions but not answered: https://github.com/Lightning-AI/lightning/discussions/14806

### How to reproduce the bug

```python
Log a metric only on rank zero and run training with `ModelCheckpoint` callback and `save_top_k=2`.
```

### Error messages and logs

```
pytorch_lightning.utilities.exceptions.MisconfigurationException: `ModelCheckpoint(monitor='val_auc')` could not find the monitored key in the returned metrics: ['train_loss', 'val_loss', 'epoch', 'step']. HINT: Did you call `log('val_auc, value)` in the `LightningModule`?
```

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):

```

**Comment by apple2373** [comment]
I also encouraged this issue. I have a metric that is defined to the entire set, and not to a batch or sub set.  So I gather all of the test output onto rank0 node, and compute the metric only on the rank0. Then this problem happened. This issue does not have the solution, so I had to do it myself.  I made the following class and solved it

```
from pytorch_lightning.utilities import rank_zero_only
class Rank0ModelCheckpoint(ModelCheckpoint):
    @rank_zero_only
    def on_validation_end(self, *args, **kwargs):
        return super().on_validation_end(*args, **kwargs)
```

then 
checkpoint_callback = Rank0ModelCheckpoint(...)

**Comment by carmocca** [comment]
Whether the checkpoint is saved on only one rank or on all ranks depends on the strategy used (DDP, DeepSpeed, FSDP, ...). As you noticed `ModelCheckpoint` logic is run for all ranks, but it does not actually save, only chooses WHEN to save.

In general, you want to log on all ranks. To work around this whilst still keeping your code, you could scatter the computed auc so that it can be logged from all ranks.

We could add logic to support what you propose by inspecting whether the metric was logged with `self.log(..., rank_zero_only=True)` (which you didn't set, that's okay because it's not used by default at the moment). But this is not implemented at the moment.

**Comment by AlessioQuercia** [comment]
Any news on this? I am also interested.

I am training on multiple GPUs with DDP and I want to validate and log on rank 0, but I also encounter the same issue.

As a workaround, I am doing the following:

``` 
@torch.no_grad()
def validation_step(self, batch, batch_idx):
    pass

@torch.no_grad()
def on_validation_epoch_end(self):
    if self.trainer.global_rank == 0:
        self._validate_and_log()
    else:
        self.trainer.callback_metrics["val/rmse"] = torch.tensor(9999999)
```

Note that the values tracked by ModelCheckpoint need to be tensors, otherwise it will trigger another error.

I am not sure if Lightning supports better solutions to achieve validation and logging on rank 0 (+ checkpoint tracking) when using multiple gpus for training with DDP. My workaround has the disadvantage that the model has to loop over the validation split of rank 0 for nothing, but it's quite fast. I also read this: https://github.com/Lightning-AI/lightning/discussions/3225.
